core/include/traccc/finding/default_detector_const_field_ckf_algorithm.hpp

```diff
@@ -0,0 +1,63 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2024 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+// Project include(s).
+#include "traccc/edm/measurement.hpp"
+#include "traccc/edm/track_candidate.hpp"
+#include "traccc/edm/track_parameters.hpp"
+#include "traccc/finding/finding_config.hpp"
+#include "traccc/geometry/detector.hpp"
+#include "traccc/utils/algorithm.hpp"
+
+// Detray include(s).
+#include <detray/detectors/bfield.hpp>
+
+namespace traccc::host {
+
+/// CKF track finding with a default Detray detector and constant magnetic field
+///
+/// This is the main host-based track finding algorithm of the project. More
+/// documentation to be written later...
+///
+class default_detector_const_field_ckf_algorithm
+    : public algorithm<track_candidate_container_types::host(
+          const default_detector::host&,
+          const detray::bfield::const_field_t::view_t&,
+          const measurement_collection_types::const_view&,
+          const bound_track_parameters_collection_types::const_view&)> {
+
+    public:
+    /// Configuration type
+    using config_type = finding_config;
+
+    /// Constructor with the algorithm's configuration
+    default_detector_const_field_ckf_algorithm(const config_type& config);
+
+    /// Execute the algorithm
+    ///
+    /// @param det The detector object
+    /// @param field The magnetic field object
+    /// @param measurements All measurements in an event
+    /// @param seeds All seeds in an event to start the track finding with
+    ///
+    /// @return A container of the found track candidates
+    ///
+    output_type operator()(
+        const default_detector::host& det,
+        const detray::bfield::const_field_t::view_t& field,
+        const measurement_collection_types::const_view& measurements,
+        const bound_track_parameters_collection_types::const_view& seeds) const;
+
+    private:
+    /// Algorithm configuration
+    config_type m_config;
+
+};  // class default_detector_ckf_algorithm
```

Comment:
I'm mostly on board, but why can't this just be a specialization of the old template type? Surely that would reduce the amount of boilerplate if we're going to end up with many different versions of this code (e.g. one constant field, one non-constant field, one combinatorial, one non-combinatorial, etc.)?

Reply:
It's a question of UI for me. :thinking: Currently I find it super hard to even just instantiate a functional finding algorithm. I have to basically copy-paste a configuration that somebody else wrote. And having users blindly copy-paste stuff is not a good setup. :frowning:

My first idea was of course just to rename `finding_algorithm.ipp` to `finding_algorithm.cpp`, instantiate a couple of concrete types in it, and be done. (Like what we do with the CUDA version at the moment.) But I just don't like that UI... :thinking:

Reply:
Okay, that's fine for me. I am happy to accept changes to these host algorithms as they will be deprecated in the not-so-distant future anyway.

But for the CUDA (and in the future SYCL) code I would like to come up with a solution that doesn't require so much boilerplate. :thinking:

Reply:
> as they will be deprecated in the not-so-distant future anyway

Not related to the discussion, but why are you going to deprecate the host algorithms? That's kinda scary

Reply:
I was under the impression that we were all on board of deprecating the traccc CPU algorithms in favour of the Acts CPU algorithms... Do you not agree? :confused:

Reply:
Not at all. You don't have to use them in ACTS or other places, but please don't delete them

Reply:
Could you motivate your point of view a bit? I'm happy to accept that you want to keep the algorithms around, sure, but I'm trying to understand why. We will need to keep this in mind when we integrate traccc into ACTS.

Reply:
They simply do their roles by:
1. Maintaining the agreement between CPU and GPU in traccc
2. Debugging any physical issue, common to CPU and GPU, in an easier way (Surprisingly, compiling and testing host algorithms is easier than doing it for GPU
3. Optimizing the tracking performance in an easier way

Most importantly they are not 100% identical to ACTS CPU algorithms, so the above things will get super difficult if they are gone.

Reply:
I am on board with adding a CMake parameter to turn off building host algorithms , if that can meet your needs

Reply:
:confused: The project's host algorithms are not going anywhere until:
  - The common CPU/GPU code sits in `ActsCore`;
  - `ActsCore` provides an equivalent of what each GPU algorithm is doing.

The design that we have in this project, where we share code between the CPU and GPU codebase, has to survive when we integrate the code into Acts. I see that as a hard requirement.

Yes, eventually the separate traccc host algorithms will disappear. But I don't see that as anything happening soon... :thinking:

Reply:
Okay, it's becoming clear to me that it would be beneficial for us all to sit down and talk about the future of traccc during either one of our bi-weekly meetings or during the ACTS developer's workshop, because I don't believe adding traccc to `ActsCore` is really on anybody's radar right now. We'll have to discuss this to produce a clear working plan.